### Borena Kunze 

## 

# Assignment Two

### Library - part two


### Task 1 and Introduction

The first task was to write code which would allow library members to borrow books. For this task it was required to create two classes "Book" and "Member", in which the method scan() would return the id of each respectively. 




In order to create a structure that would be as clear as possible I decided at first to create three classes "Book", "Member" and "Lending" each containing the specific attributes as well as the class "Library", which would provide the functionalities needed for each task.   
Alongside with the classes I have also read in the three csv files, the book.csv, the members.csv and the bookloans.csv which would correspond to the classes.  
(As the main work with the csv files was part of assignment one, I have kept the reading of the files simple to provide better readability for the code, which is more complex and intertwined in this assignment.)  



Later for for the task 5.2 I have also added the class "OrderedBook" which inherits from the class Book and would make it possible to notify a member that an ordered book has become available.  
Wherever possible I have tried to add testing of the code to demonstrate that all parts of the code are working.


The main challenge for this assignment is surely to provide an understandable and clear structure alongside functionality. As the main code is included in the Library class, I have kept the testing wherever possible at the individual section of each of the taks.

In [1]:
#importing libraries

import csv
import datetime
import json

--- 
## Book List

In [2]:
#creates the class book

class Book:
    '''
    The class Book stores the attributes number, title, author, genre, subgenre, publisher corresponding to those given in the 
    file books.csv. 
    The method scan() returns the id "number" of a specific book
    '''
    def __init__(self, number, title, author, genre, subgenre, publisher):
        self.number = number
        self.title = title
        self.author = author
        self.genre = genre
        self.subgenre = subgenre
        self.publisher = publisher
    
    def scan(self):
        '''
        scan method returns the number id of the book
        '''
        return self.number
        

class OrderedBook(Book):      #task 5 subtask 2
    '''
    The class OrderedBook, created for Task 5.2 to make it possible to notify a member that an ordered book has become available, 
    inherits from the Book class and adds the attribute memberId in order to connect the ordered book to the member who has ordered it.
    '''
    def __init__(self,number,title,author,genre,subgenre,publisher, memberId):
        super().__init__(number,title,author,genre,subgenre,publisher)
        self.memberId = memberId        #adds attribute
        
        
        
'''
Testing Book class and scan() method
'''

newbook = Book('1','Fundamentals of Wavelets',"Goswami, Jaideva",'tech','signal_processing','Wiley')   #creating the object
print("The title of the book is : {}".format(newbook.title))
print("The genre of the book is : {}".format(newbook.genre))     #object.attribute
print("The id of the book is : {}".format(newbook.scan()))       #object.method 


The title of the book is : Fundamentals of Wavelets
The genre of the book is : tech
The id of the book is : 1


The book class containing the book attributes is the first class which I created. It also contains the suggested scan function which returns the number id of each book. The functionality is tested at the bottom of the cell with data from the csv file, showing attributes of the book and the method scan.   


The class OrderedBook which serves task 5.2 makes it possible to notify a member when the book that he/she has ordered is available for loaning. OrderedBook inherits from the class Book adding the attribute memberId. 

The next step would be to read in the book.csv file. To keep the code as clear as possible I have decided to create lists which will also allow me to iterate and work accross them more easily in the later tasks. The book csv file contains information on 120 books, encoded as UTF-8, each with a given unique number and attributes such as book title, author, genre etc. 
To avoid it taking too much space, I decided to not print it out, as it is not necessary at this point. Try, except, finally are used for error handling.

In [3]:

'''
Reading in the books.csv file and creating a list of books storing the information of the csv file
Using try, except, finally for error handling
'''

try:                                                                     #reading the file and
    with open('books.csv', mode = 'r', encoding = 'utf-8') as file:      #using 'utf-8' to encode the file
        csvFile = csv.reader(file)                                       
        #print(csvFile)
        
        rowCount = 0
        bookList = []            #creating the list of books
        
        for row in csvFile:
            #print(row)
            if rowCount == 0:
                #reads the heading
                heading = row        
                #print(heading)
                
                rowCount += 1
            else:
                #read and store the content 
                bookId = row[0]
                title = row[1]
                author = row[2]
                genre = row[3]
                subgenre = row[4]
                publisher = row[5]
                
                newbook = Book(bookId, title, author, genre, subgenre, publisher)
                bookList.append(newbook)
                
            
except:
    print('File could not be opened.')
    
finally:
    file.close()

--- 
## Members

To create the class Member I followed the same procedure as for the Book class, creating a Member class which contains all the member's attributes, such as id, first and last name, gender, email and card number as well as the scan method which returns here the member id "ID".  
To test the functionality print examples of attributes as well as the scan method are provided at the end of the cell. 

In [4]:
#creating class member

class Member:
    '''
    Member class contains the attributes such as ID, first_name, last_name, gender, email and cardnumber
    The method scan() returns the member ID
    
    '''
    def __init__(self, ID, first_name, last_name, gender, email, cardnumber):
        self.ID = ID
        self.first_name = first_name
        self.last_name = last_name
        self.gender = gender
        self.email = email
        self.cardnumber = cardnumber
    
    def __str__(self):                       #avoids printing the object position in memory and prints the actual information
        return self.first_name + " " + self.last_name                        #returns name and surname of the member
        
    def scan(self):
        '''
        scan method returns member ID
        '''
        return self.ID           

    
'''
Printing examples to ensure that the code is working
'''
member = Member('1','Adelaide','Cunningham','Female','a.cunningham@randatmail.com','13')     #creating object
print("The email address of the member is: {}".format(member.email))
print("The id number of the member is: {}".format(member.scan()))      #testing scan() method
print("The new member is: {}".format(member))


The email address of the member is: a.cunningham@randatmail.com
The id number of the member is: 1
The new member is: Adelaide Cunningham


In [5]:
#reading the members.csv file
#creating the membersList list

'''
Reading in the members.csv file and creating a list of members storing the information from the csv file
Using 'utf-8' for encoding and
try, except, finally for error handling
'''

try:
    with open('members.csv', mode = 'r', encoding = 'utf-8') as file:      #using 'utf-8' to encode the file
        csvFile = csv.reader(file)
        #print(csvFile)
        
        rowCount = 0
        membersList = []
        
        for row in csvFile:
            print(row)
            if rowCount == 0:
                #reads the heading
                heading = row        
                #print(heading)
                
                rowCount += 1
            else:
                #read and store the content in the dictionary
                memberId = row[0]
                firstName = row[1]
                lastName = row[2]
                gender = row[3]
                email = row[4]
                cardnumber = row[5]
                
                newMember = Member(memberId, firstName, lastName, gender, email, cardnumber)
                membersList.append(newMember)
                
            
except:
    print('File could not be opened.')
    
finally:
    file.close()
    

['\ufeffID', 'First Name', 'Last Name', 'Gender', 'Email', 'CardNumber']
['1', 'Adelaide', 'Cunningham', 'Female', 'a.cunningham@randatmail.com', '13']
['2', 'Charlie', 'Roberts', 'Male', 'c.roberts@randatmail.com', '22']
['3', 'Eric', 'Cooper', 'Male', 'e.cooper@randatmail.com', '33']
['4', 'Cadie', 'Hall', 'Female', 'c.hall@randatmail.com', '43']
['5', 'Darcy', 'Howard', 'Female', 'd.howard@randatmail.com', '52']
['6', 'Connie', 'West', 'Female', 'c.west@randatmail.com', '61']
['7', 'Lyndon', 'Ellis', 'Male', 'l.ellis@randatmail.com', '72']
['8', 'Amy', 'Hamilton', 'Female', 'a.hamilton@randatmail.com', '82']
['9', 'Kelvin', 'Wilson', 'Male', 'k.wilson@randatmail.com', '93']
['10', 'Arianna', 'Watson', 'Female', 'a.watson@randatmail.com', '0']
['11', 'Lydia', 'Farrell', 'Female', 'l.farrell@randatmail.com', '113']
['12', 'Ada', 'Harris', 'Female', 'a.harris@randatmail.com', '122']
['13', 'Sarah', 'Wilson', 'Female', 's.wilson@randatmail.com', '131']
['14', 'Nicole', 'Phillips', 'Fema

As this file was not part of the first assignment I have printed the content out, creating the opportunity to become familiar with the information included. There are 200 members listed, meaning that any new member will have an id greater than 200 and their membership cards will be from 2011 onwards. This will be of interest to provide the possibility of adding new members and creating the membership cards for them.

--- 
## Lending Books

The class Lending will provide functionality for lending a book. Book Id, member Id, the date of loan as well as the date of return are the stored attributes and will be used later to calculate the books which are being returned late and allow to send the notifications.   
The printout expresses the status of the loaned book showing that it is on loan and the member who has borrowed it.

In [6]:
#class for lending books

class Lending:
    '''
    Lending contains the attributes of the provided bookloans information, such as book_ID, member_ID, date_of_loan, date_of_return
    The converting of the date format will be provided through the function formatDate in the following cell.
    '''
    def __init__(self, bookID, memberID, date_of_loan, date_of_return):
        self.book_ID = bookID
        self.member_ID = memberID
        self.date_of_loan = date_of_loan
        self.date_of_return = date_of_return

'''
Testing the class Lending with given information
'''
lend = Lending('1','71','43726','43746')
print("The book with the id {} is currently loaned by member {}.".format(lend.book_ID, lend.member_ID))

The book with the id 1 is currently loaned by member 71.


To allow for further calculations, the time format needed to be converted into a human readable format, such as Year-Month-Day. Using datetime I have created a function which returns the given dates from the Microsoft epoch format into a year-month-day format. 


In [7]:
#convert the numerical data to dates


def formatDate(refValue, dateValue):
    '''
    Function converts Microsoft epoch formatted dates 
    return: readable date as Year-Month-Day 
    '''
    #considering epoch date to be 1900-01-01
    formatRef = datetime.datetime.strptime(refValue, "%Y-%m-%d")                         #converting the date into readable format
    newDate = (formatRef + datetime.timedelta(days=int(dateValue))).strftime("%Y-%m-%d")
    return newDate              #returns the date in the new format




As with the previous files, the following code will read in the bookloans.csv file into the list loanList whilst applying the formatDate function created previously to convert the dates. 
At the same time the rows with a book id greater than 120 will be excluded as they do not correspond with the list of books containing only 120 book entries. Again try, except, finally is used for error handling and tests are provided at the end showing when the book was loaned and when it is returned.


In [8]:

'''
Opening and reading in the bookloans.csv file 
Excluding all rows where the unique book id is > 120 
Excluding the rows where the return date is 0
Applying the formatDate function
Using try, except, finally for error handling
'''
# open the file bookloans.csv
try:
    #opening file
    with open('bookloans.csv', mode = 'r', encoding = 'utf-8') as file:       #using 'utf-8' to encode
        csvFile = csv.reader(file)
        #print(csvFile)

        #initialize the list of loans
        loanList = []


        #read the data
        for row in csvFile:
            #changing first element into '1'
            if row[0].isdigit() == False:           
                row[0] = '1'
                # excluding elements where return_date is 0
                # excluding elements where the bookId > 120
            if row[3] != 0 and int(row[0]) < 121:

                if row[0] != 0:                           #reading in rows 0 and 1
                    bookID = row[0]
                if row[1] != 0:
                    memberID = row[1]
                if row[2] != 0:
                #changing epoch into dates
                    date_of_loan = formatDate('1900-01-01', row[2])   #append the formatDate function
                else:
                    date_of_loan = 'Nan'
                if row[3] != 0:
                    date_of_return = formatDate('1900-01-01', row[3])
                else:
                    date_of_return = 'Nan'

                #create object 
                lending = Lending(bookID, memberID, date_of_loan, date_of_return)

                loanList.append(lending)


        
except:
    print('File could not be opened.')
    
finally:
    file.close()


'''
Testing loanlist
'''
print("The book with the id {} is loaned by member {} on the {} and has been returned on {}.".format(loanList[4].book_ID, loanList[4].member_ID, loanList[4].date_of_loan, loanList[4].date_of_return))
print("The book with the id {} is loaned by member {} on the {} and has been returned on {}.".format(loanList[43].book_ID, loanList[43].member_ID, loanList[43].date_of_loan, loanList[43].date_of_return))


The book with the id 1 is loaned by member 38 on the 2019-03-21 and has been returned on 2019-03-26.
The book with the id 3 is loaned by member 133 on the 2019-01-06 and has been returned on 2019-01-23.


--- 
## Library

After reading in all three csv files and creating the three classes, the class Library will provide functionality needed to solve the different tasks.   
The Library class includes therefore various methods belonging to different tasks, providing a simple and unified structure for the library management whilst making it possible to add further functionalities when needed.   
I have added explanations and comments wherever possible for better understanding.  

The following explanations should provide further clarifications on the different functionalities within the Library class:

The methods hasMember and hasBook use the method scan to check if a book or a member are registered in the library, returning True if they are and False otherwise.  




The method loan is concerned with the different aspects of loaning a book. After checking that the provided member id is registered and the book code is available, it will allow for the book to be loaned and registered as unavailable for loaning.  
If a book appears as unavailable, it will offer the opportunity to reserve by adding it to the reservation_list alongside the id of the member requesting it until it becomes available. (Task 4) Further in the method bookReservation the reservation_list is stored into a dictionary and finally printed out as a json file. To complete the requirements of Task 5, the method sendEmailReturnedBook provides the functionality to send a notification to the member who has reserved the book when the book has again become available for loaning.  

When the provided book code is not registered in the library, the option to place a book order is offered. (Task 5.2) The member requesting the book will be able to enter the book details such as title, author etc, so that the book object can be added to the list of ordered books orderedList. Later the notifiedOrder method would send an email to the member who has ordered the book by extracting his id and email from members and printing the message saying that the book has become available. (Task 5.2)

To ensure that a loaned book can be returned (Task 2) after checking that both member id and book id are valid entries, the method returnBook will overwrite the '1900' date which indicates that a book is not returned yet, to the current from date new_return_date.

Regarding Task 3 for which the addition of new members was required, I have added the createNewMember method, which takes the first and last name, gender and email of the new member as an input. Based on the number of current members it generates a new id and a new card number for the new member. All information of the new member is then added to the list of current members as well as to todaysList, which is used to create the dictionary  newMemberData and writing a json file which would then be sent to the external company to produce the membership cards.

The last method fineNotification, which takes the date set as todaysDate as an input calculates if a book is being returned late and has been loaned for longer than 14 days. In this case an email is sent to the member who is currently loaning the book.

In [9]:
class Library():
    
    def __init__(self, list_of_books=[], list_of_members=[], loanList=[]):
        """
        Defines a Library class with a list of book (default empty) and list of members (default empty)
        
        """
        # protection
        if len(list_of_books) != 0:
            for el in list_of_books:
                assert isinstance(el,Book)   #checking the types of the elements of the list, they have to be of the type book
        if len(list_of_members) != 0:
            for el in list_of_members:       #the same for the list of members
                assert isinstance(el,Member)        
        self._books = list_of_books
        self.members = list_of_members      
        self.loanList = loanList            #will be used for loaning books
        self.reservation_list = []          #will be used for task 4 Reserving a book
        self.orderedList = []               #for task 5.2 Ordering a book
        
        self._state_of_books = {}
        for el in loanList:
            if el.date_of_return[0:4] == '1900':
                self._state_of_books[el.book_ID] = False        #False if the book is loaned
            else:
                self._state_of_books[el.book_ID] = True         #True if available
            
    
    def hasMember(self, code):            
        '''
        Checks with the provided code if the holder is a member of the library
        by using the scan() method
        returns True if code is found
        '''
        for member in self.members:
            if member.scan() == code:                      
                return True
        return False
    
    def hasBook(self, code):
        '''
        Checks with the provided book code if the book is available in the library
        by using the scan() method
        returns True if book code is found
        '''
        for book in self._books:
            if book.scan() == code:
                return True
        return False
    
    def addBook(self, book):
        '''
        Sets book value to "True" for a book id in state_of_books dict, when it is available for loan
        '''
        self._books.append(book)
        self._state_of_books[book.scan()] = True
        
    def loan(self, book_code, member_code, date_of_loan='', date_of_return=''):
        '''
        Method to allow a book to be loaned
        input: book_code, member_code, date_of_loan, date_of_return
        
        Checks: if a person is a member and if the requested book is available in the library
        If the person is a member and the book is available, the bookloan object created with Lending
        containing book_code, member_code, date_of_loan, date_of_return will be added to the loanList and therefore 
        loaning the book will be recorded
        
        If the book is currently on loan not available for loaning, the option to reserve the book will be offered 
        and book and member codes will be added to the reservation_list
        
        In case that the provided book code is not a registered code of the library books, the option to order the
        book is offered, and the details of the book are requested

        '''
        if self.hasMember(member_code): # check that the person is a member
            # allow the loaning
            if self.hasBook(book_code): # check that the library owns the book
                print("The book code is: {}".format(book_code))
                print("The book is now available: ", self._state_of_books[book_code])
                if self._state_of_books[book_code]:
                    self._state_of_books[book_code] = False    #False as we are giving the book away, therefore its "status" will change to False in state_of_books
                    bookloan = Lending(book_code, member_code, date_of_loan, date_of_return) #creates object
                    self.loanList.append(bookloan)      #adds a loaned book to the library
                    #print("The book {} is loaned to the member {}. The loaning date was {} and the return date is {}".format(book_code, member_code, date_of_loan, date_of_return))
                    return True     #True when loaning is done
                else:
                    print("The book is currently unavailable for loaning")      #if book is currently not available
                    reserving = input("Do you want to reserve the book? y/n")   #task 4 #Reservation of book
                    if reserving == "y":
                        book = {}            #create dictionary taks 4
                        book["Book"] = book_code
                        book["Member"] = member_code
                        self.reservation_list.append(book)
                    print("The book {} is loaned to the member {}. The loaning date was {} and the return date is {}".format(book_code, member_code, date_of_loan, date_of_return))
                    return False   #False if loaning was cancelled 
            else:
                print("The library does not own Book with code {}".format(book_code))  #Task 5.2 ordering the book
                answer = input("Do you want to reserve the book? y/n")
                if answer == 'y':
                    title = input("Please enter the title ")
                    author = input("Please enter the author ")
                    genre = input("Please enter the genre ")
                    subgenre = input("Please enter the subgenre ")
                    publisher = input("Please enter the publisher ")
                    orderedBook = OrderedBook(book_code, title, author, genre, subgenre, publisher, member_code)
                    self.orderedList.append(orderedBook)       #orderedBook object is added to list of ordered books orderedList
                return False
        else:
            print("Person with code {} is not a member of this library".format(member_code))
            return False
        
    def returnBook(self, book_code, member_code, new_return_date):     #this is about task 2 Returning a book
        '''
        Method providing functionality to return a book 
        Input: book_code, member_code, new_return_date
        In case that the provided member and book id are registered with the library and the return date is 1900
        date_of_return will be overwritten with the new date new_return_date (Testing in Task 2)
        
        In addition as required in Task 5, an email will be sent to the member who has reserved the book
        '''
        if self.hasMember(member_code):
            if self.hasBook(book_code):
                for element in self.loanList:
                    if element.member_ID == member_code and element.book_ID == book_code:
                        print("The book requested with the number {} is found".format(book_code))
                        print("The book was loaned on {} and returned on {}.".format(element.date_of_loan, element.date_of_return))
                        if element.date_of_return[0:4] == '1900':      #is not returned yet
                            element.date_of_return = new_return_date   #overwrites the 1900 date (which shows that it is not returned)
                            #print("The book was loaned on {} and returned on {}.".format(element.date_of_loan, element.date_of_return)) 
                            self.sendEmailReturnedBook(book_code)      #task 5 Send notification to member who has reserved the book
            else:
                print("The library does not own Book with code {}".format(book_code))    
        else:
            print("Person with code {} is not a member of this library".format(member_code))
            
            
    def createNewMember(self, firstName, lastName, gender, email):            #Task 3 Adding new members  
        '''
        Method to allow the registration of a new member 
        Input: firstName, lastName, gender, email
        Creates the id number for the new member based on the number of the current members
        Creates the cardnumber for the new member
        '''
        number_of_members = len(self.members)
        ID = str(number_of_members + 1)                  #creates the id number 
        cardnumber = ID + str(1)                         #creates the card number
        new_member = Member(ID, firstName, lastName, gender, email, cardnumber) #object new_member 
        #print(cardnumber, ID)
        self.addMember(new_member)
        
    def addMember(self, member):
        '''
        Method to assign member to members (in Library)
        '''
        self.members.append(member)          
        
        
    def todaysMembers(self, member, todaysList):  
        '''
        Method to add new members to the daily list todaysList
        returns todaysList
        '''
        todaysList.append(member)
        return todaysList
    
    
    def newMembersFileJSON(self, filename, todaysList):
        '''
        Method to create dictionary containing the new member's information
        and writing a json file
        '''
        newMemberData = {}
        newMemberData["people"] = []                #key people and value list
        for i in range(len(todaysList)):            #creates dictionary with the list of information 
            newMemberData["people"].append({        #iterates through todaysList adding the member info to the dict
                "ID": todaysList[i].ID,
                "firstName": todaysList[i].first_name,
                "lastName" : todaysList[i].last_name,
                "gender" : todaysList[i].gender,
                "email" : todaysList[i].email,
                "cardnumber" : todaysList[i].cardnumber
            })
        
        writeFile = filename + ".json" 
        with open(writeFile, 'w') as file:
            json.dump(newMemberData, file)              #writing the json file for new members
        
    
    def bookReservation(self, filename):                 #creating the json file with the reservations
        '''
        Method to create the reservation_dict containing reservation info
        Writes a json file from the reservation dict
        '''
        reservation_dict = {}
        reservation_dict["Reserved"] = self.reservation_list
        writeFile = filename + ".json"
        with open(writeFile, 'w') as file:               #writing the json file
            json.dump(reservation_dict, file)

        '''
        The following methods concerning task 5 
        Subtask 1: Notification that a reserved book had become available.
        Subtask 2: Notification that an ordered book had become available.
        Subtask 3: Notification that a book return was late and of a resultant fine.
        '''

        
    def sendEmailReturnedBook(self, book_code):    #task 5.1
        '''
        Sends a notification to the member when the reserved book is available
        '''
        for reservation in self.reservation_list:
            if reservation["Book"] == book_code:
                member_code = reservation["Member"]
                for member in self.members:
                    if member.ID == member_code:
                        email = member.email
                        print("The book with the id {} is now available for {}. Sending an email to {}".format(book_code, member_code, email))
   

    def notifiedOrder(self, arrivedBooks = []):  #task 5.2
        '''
        Sends notification to the specific member who has ordered the book when it is available
        '''
        for arrivedBook in arrivedBooks:
            for orderedBook in self.orderedList:
                if orderedBook.number == arrivedBook:
                    for member in self.members:
                        if member.ID == orderedBook.memberId:
                            print("The book {} ordered by {} has arrived. Sending an email to {}". format(arrivedBook, member.ID, member.email))
    
    
    def fineNotification(self, todaysDate):      #task 5.3
        '''
        Sends a fine if the book loaned is not returned after 14 days and to the books which are on loan
        longer than 14 days
        '''
        for element in self.loanList:                   #iterates through loanList
            if element.date_of_return[0:4] == '1900':   #year '1900' means that a book is not returned
                loanDate = datetime.datetime.fromisoformat(element.date_of_loan)
                today = datetime.datetime.fromisoformat(todaysDate)            #today
                if (today - loanDate).days > 14:        #calculates if more than 14 days have passed
                    memberId = element.member_ID
                    for member in self.members:         #extracts member id and email information
                        if member.ID == memberId:
                            email = member.email
                            print("The member {} has not returned the book {}. A fine will be sent to {}".format(memberId, element.book_ID, email))
    


### Task 1 : Loaning a book

The following example demonstrates the functionality of the code in the situation where a book is requested but is not available for loaning. The member is offered the opportunity to make a reservation by entering y for yes, or to refrain by entering n for no.

In [10]:
#Book is unavailable for loaning and will be reserved

library = Library(bookList, membersList, loanList)   #create library object

print(library.loanList[-1].member_ID)

library.loan('2', '104', '2021-04-19', '2021-05-19') #adds a new loaned book
print(library.loanList[-1].member_ID)



1
The book code is: 2
The book is now available:  False
The book is currently unavailable for loaning
Do you want to reserve the book? y/ny
The book 2 is loaned to the member 104. The loaning date was 2021-04-19 and the return date is 2021-05-19
1


In this situation the book is available for loaning and therefore the loan is successful. The printout shows the  id of the book which is being loaned and the member id of the member loaning the book. The provided dummy dates represent the date of loan and the date of return.

In [11]:
#Book is available for loaning

library = Library(bookList, membersList, loanList)   #create library object

print(library.loanList[-1].member_ID)

library.loan('3', '85', '2021-04-19', '2021-05-19') #adds a new loaned book with dummy dates
print(library.loanList[-1].member_ID)


1
The book code is: 3
The book is now available:  True
85


The following code is testing the loan of a book by firstly loaning the book to a member and then requesting the same book again.  
When requesting the book a second time the member is offered the option to reserve the book. This shows that the loaning was successful the first time and the book has been recorded as unavailable.  

In [12]:
newbook = Book('1','Fundamentals of Wavelets',"Goswami, Jaideva",'tech','signal_processing','Wiley')

In [13]:
# Test the loan of a book

if library.loan(newbook.scan(), newMember.scan()) == True:     
    print("Book", newbook.title, "loaned to", newMember.first_name)
else:
    print("Book", newbook.title, "not loaned to", newMember.first_name)
    

The book code is: 1
The book is now available:  True
Book Fundamentals of Wavelets loaned to Marcus


In [14]:
newbook = Book('1','Fundamentals of Wavelets',"Goswami, Jaideva",'tech','signal_processing','Wiley')

In [17]:
#Test the loan when the book is already borrowed

if library.loan(newbook.scan(), newMember.scan()) == True:     
    print("Book", newbook.title, "loaned to", newMember.first_name)
else:
    print("Book", newbook.title, "not loaned to", newMember.first_name)
    

The book code is: 1
The book is now available:  True
Book Fundamentals of Wavelets loaned to Marcus


### Task 2 : Returning a book 

The following code tests the returning of a book with a given "dummy" date of return and the existing valid cardnumber and book id.

In [18]:
#Testing functionality with bookloans.csv file entry: '2','18','43830','0'


library = Library(bookList, membersList, loanList)

cardnumber = '18'

booknumber = '2'

date_of_return = '2021-04-30'

library.returnBook(booknumber, cardnumber, date_of_return)   



The book requested with the number 2 is found
The book was loaned on 2019-02-15 and returned on 2019-02-17.
The book requested with the number 2 is found
The book was loaned on 2020-01-02 and returned on 2021-04-30.


"The book was loaned on 2020-01-02 and returned on 2021-04-30."
shows that the '0' entry in date_of_return has been successfully updated into the current given date '2021-04-30', which means that the book will be available when a new loan request is made.

### Task 3 : Adding new members

As requested in task 3 the code allows for new members to be added and their details to be stored. I have added two new fictional members as an example. To demonstrate the functionality I am printing first the last already existing library member. Then two new members are added with newly created ids and cardnumbers. When the last member is printed out again, it will show the last entry as the last library member.

It would be best to not run the code more than once as the member number will be updated again. 

In [19]:
#Creating daily list of new members
#Code is tested on two examples
#If this code is run more than once it will just add the same two members and update their id and cardnumbers

todaysList = []
library = Library(bookList, membersList, loanList)

#print the last existing member, the cardnumber and id before adding the new members
print(library.members[-1].first_name, library.members[-1].cardnumber, library.members[-1].ID)  

library.createNewMember("Wilfred", "Mott", "male", "wmott@example.com") #add new member

print(library.members[-1].first_name, library.members[-1].cardnumber, library.members[-1].ID) #prints added member

todaysList = library.todaysMembers(library.members[-1], todaysList)

library.createNewMember("Sarah Jane", "Smith", "Female", "sarahjsmith@example.com") #add new member

#print last member after adding two new members
print(library.members[-1].first_name, library.members[-1].cardnumber, library.members[-1].ID) 

todaysList = library.todaysMembers(library.members[-1], todaysList)
print("Today {} new members have been added.".format(len(todaysList))) #prints how many new members have been added


Marcus 2002 200
Wilfred 2011 201
Sarah Jane 2021 202
Today 2 new members have been added.


Running the following cell will ensure that the a json file "daily_new_members.json" containing the information on the new members is created and saved in the current directory. This file could then be sent to the company which prints the cards for the new members. 

In [20]:
#new members are stored in a json file

filename = "daily_new_members.json"

library.newMembersFileJSON('daily_new_members', todaysList)

The json file containing the information of each new member as well as their id and generated cardnumber would store the information as follows:  

{"people": [{"ID": "201", "firstName": "Wilfred", "lastName": "Mott", "gender": "male", "email": "wmott@example.com", "cardnumber": "2011"}, {"ID": "202", "firstName": "Sarah Jane", "lastName": "Smith", "gender": "Female", "email": "sarahjsmith@example.com", "cardnumber": "2021"}]}

### Task 4 : Reserving a book

Task: Provide functionality to allow a member to reserve a book. Assume that they have the number of the book available. This will require permanent storage as it may be some time before a book becomes available.

In task four I am creating a json file in which book reservation requests are stored. Based on the book code provided a member can request a book and can choose to make a reservation if the book is not available at that moment by responding with 'y' for yes. There is no actual limit set on how many books a member can request at a time, as it did not seem necessary. However this is something that could be implemented too if needed.    

The file Reservations.json containing the book id together with the member id will be stored in the current directory as follows.  


{"Reserved": [{"Book": "106", "Member": "30"}, {"Book": "105", "Member": "30"}]}  


In [21]:
#task 4

library = Library(bookList, membersList, loanList)     

library.loan('106', '30')      # Member 30 requesting book 106 and 105
library.loan('105', '30')
library.bookReservation('Reservations')   #file Reservations.json is created

The book code is: 106
The book is now available:  False
The book is currently unavailable for loaning
Do you want to reserve the book? y/ny
The book 106 is loaned to the member 30. The loaning date was  and the return date is 
The book code is: 105
The book is now available:  False
The book is currently unavailable for loaning
Do you want to reserve the book? y/ny
The book 105 is loaned to the member 30. The loaning date was  and the return date is 


### Task 5 : Sending notifications

Task: Test the system and show or describe how it is capable of coping with the situations described in the preamble and how it might be sufficiently flexible to deal with future notification requirements.

1. Notification that a reserved book had become available.  
2. Notification that an ordered book had become available.  
3. Notification that a book return was late and of a resultant fine.

The following code tests the functionality of Task 5.1. It creates a reservation for a book that is currently not available and sends an email to the member who has requested it.  
In a real life situation there would be a time gap between reserving the book and it becoming available again. However in the test situation this should not be a concern.

In [75]:
library = Library(bookList, membersList, loanList)   #library object

#Task 5.1 Notification that a reserved book had become available.

library.loan('8', '167')                        #book is reserved for member '167'

library.returnBook('8', '167', '2021-04-30')    #book is returned


The book code is: 8
The book is now available:  False
The book is currently unavailable for loaning
Do you want to reserve the book? y/ny
The book requested with the number 8 is found
The book with the id 8 is now available for 167. Sending an email to a.ferguson@randatmail.com


The following code allows the member to order a book if it is not available in the library. The member needs to provide the necessary details of the book, such as title, author etc and the order will be placed. As soon as the book arrives the member will be notified via email.  
I have tested the code with two different examples showing that different members can order different books at the same time. 

The code could be changed to entering the book title first to allow for easier recognition of the book title in a real situation, however for the purpose of this task I felt that the book id was sufficient.

In [76]:
'''
Testing notification by ordering two books and sending the notification tho the person who has placed the order
'''
#Task 5.2 Ordering a book

library.loan('203', '29')


The library does not own Book with code 203
Do you want to reserve the book? y/ny
Please enter the title Macbeth
Please enter the author William Shakespeare
Please enter the genre dramatic play
Please enter the subgenre tragedy
Please enter the publisher Wordsworth Editions


False

In [77]:
#Task 5.2 Ordering a second book

library.loan('204', '15')

The library does not own Book with code 204
Do you want to reserve the book? y/ny
Please enter the title The Castle
Please enter the author Franz Kafka
Please enter the genre fiction
Please enter the subgenre modernist fiction
Please enter the publisher Oxford World's Classics


False

Both members will be notified separately when the books they have ordered are available. The printout mentions the book id and the member id as well as the member's email address. As this process would be automated I think it is less error prone to work with just the ids for this task.

In [79]:
#Task 5.2 Sending a notification when an ordered book has arrived

library.notifiedOrder(['203'])

The book 203 ordered by 29 has arrived. Sending an email to m.miller@randatmail.com


In [78]:
#Sending a different notification when an ordered book has arrived to the member who requested the book

library.notifiedOrder(['204'])

The book 204 ordered by 15 has arrived. Sending an email to m.anderson@randatmail.com


Task 5.3 required to send notifications to all members who did not return the books in time. The following code will calculate which of the books currently on loan have been kept for longer than 14 days based on todaysDate and the given date '2021-04-30'. As I have chosen a more recent date there are a few fines issued.  

In [74]:
#Task 5.3 Sending fines to all books kept longer than 14 days with a given date '2021-04-30'

library.fineNotification('2021-04-30')



The member 120 has not returned the book 6. A fine will be sent to e.douglas@randatmail.com
The member 167 has not returned the book 8. A fine will be sent to a.ferguson@randatmail.com
The member 118 has not returned the book 15. A fine will be sent to a.carter@randatmail.com
The member 42 has not returned the book 17. A fine will be sent to a.harris@randatmail.com
The member 69 has not returned the book 22. A fine will be sent to a.cooper@randatmail.com
The member 78 has not returned the book 24. A fine will be sent to l.richards@randatmail.com
The member 23 has not returned the book 30. A fine will be sent to a.murray@randatmail.com
The member 34 has not returned the book 31. A fine will be sent to c.phillips@randatmail.com
The member 45 has not returned the book 32. A fine will be sent to r.montgomery@randatmail.com
The member 168 has not returned the book 33. A fine will be sent to j.davis@randatmail.com
The member 30 has not returned the book 36. A fine will be sent to p.elliott@r

Overall the library management system created so far is providing the basic functionalities to register members, books, to loan or reserve a book and to send notifications. Further functionalities could be added if needed in the future and the given functionalities could be adjusted to the specific needs of the library.   
This could e.g. include refining the notification system, so that it would send different timed reminder messages to the members before issuing the final fine. 
Testing of the system is provided wherever possible to demonstrate the different steps and to show how the tasks set are solved. 


Thank you.